internal/yunikorn/sync.go

```diff
@@ -260,55 +251,54 @@ func (s *Service) upsertPartitionNodes(ctx context.Context, partitions []*dao.Pa
 	return nil
 }
 
-// upsertApplications fetches applications for each queue and upserts them into the database
-func (s *Service) upsertApplications(ctx context.Context, queues []*dao.PartitionQueueDAOInfo) error {
+// syncApplications fetches applications for each queue and upserts them into the database
+func (s *Service) syncApplications(ctx context.Context) error {
 	logger := log.FromContext(ctx)
 
-	// Create a wait group as a separate goroutine will be spawned for each partition
-	wg := sync.WaitGroup{}
-	wg.Add(len(queues))
-
-	// Protect the applications and errors slices with a mutex so multiple goroutines can safely append queues
-	mutex := sync.Mutex{}
-
-	var errs []error
-	var apps []*dao.ApplicationDAOInfo
+	applications, err := s.client.GetApplications(ctx, "", "")
+	if err != nil {
+		return fmt.Errorf("could not get applications: %v", err)
+	}
 
-	processQueue := func(q *dao.PartitionQueueDAOInfo) {
-		defer wg.Done()
-		queueApps, err := s.client.GetApplications(ctx, q.Partition, q.QueueName)
-		if err != nil {
-			mutex.Lock()
-			errs = append(
-				errs,
-				fmt.Errorf("could not get applications for partition %s, queue %s: %v", q.Partition, q.QueueName, err),
-			)
-			mutex.Unlock()
-		} else {
-			mutex.Lock()
-			apps = append(apps, queueApps...)
-			mutex.Unlock()
-		}
+	dbApplications, err := s.repo.GetLatestApplicationsByApplicationID(ctx)
+	if err != nil {
+		return fmt.Errorf("could not get latest applications: %v", err)
 	}
 
-	// Get applications from Yunikorn API and upsert into DB
-	for _, q := range queues {
-		go processQueue(q)
+	lookup := make(map[string]*model.Application, len(dbApplications))
+	for _, a := range dbApplications {
+		lookup[a.ApplicationID] = a
 	}
 
-	// wait for all queues to be processed
-	wg.Wait()
+	now := time.Now().UnixNano()
+	for _, a := range applications {
+		current, ok := lookup[a.ApplicationID]
+		delete(lookup, a.ApplicationID)
+		if !ok || current.DeletedAt != nil { // either not exists or deleted
+			application := &model.Application{
+				ModelMetadata: model.ModelMetadata{
+					ID:        ulid.Make().String(),
```

Comment:
I'd suggest having a `NewModelMetadata` which has this logic for creating IDs instead of manually instantiating it, but that is up to you